File: ui/src/utils/device-fragment.ts
import { isMap, isSeq, type Node, parseDocument } from 'yaml';

/**
 * A single device's slice of the whole-config YAML.
 *
 * The detail pane edits one device, but the daemon only accepts the whole
 * config, so the edit has to be put back where it came from. Doing that
 * through a YAML round-trip would reformat the entire file and drop the
 * comments operators write in it, so the fragment is a *byte range* instead:
 * everything outside `[start, end)` is copied through untouched.
 */
export interface DeviceFragment {
  /** The device's YAML, dedented so the editor shows a standalone document. */
  text: string;
  /** Offset of the first character of the device's block (its `- ` marker). */
  start: number;
  /** Offset just past the device's block. */
  end: number;
  /** Indent the list items sit at, restored on splice. */
  indent: string;
}

/** Offset of the start of the line containing offset. */
function lineStart(source: string, offset: number): number {
  const previous = source.lastIndexOf('\n', offset - 1);
  return previous + 1;
}

/**
 * Offset just past the device's last content line. The parser's value-end can
 * run past trailing blank lines; those belong to the gap between devices, not
 * to the device, so they are excluded — otherwise editing a device silently
 * closes up the spacing around it.
 */
function blockEnd(source: string, start: number, valueEnd: number): number {
  const nextNewline = source.indexOf('\n', valueEnd);
  const end = nextNewline === -1 ? source.length : nextNewline + 1;
  return trimTrailingBlankLines(source, start, end);
}

/** Start of the line the next sequence item begins on, or null when this is
 * the last item. */
function nextItemStart(source: string, next: unknown): number | null {
  const range = (next as Node | undefined)?.range;
  return range ? lineStart(source, range[0]) : null;
}

/** Walks `end` back over blank lines, so the gap between two devices belongs
 * to neither of them and editing one does not close up the spacing. */
function trimTrailingBlankLines(source: string, start: number, end: number): number {
  let trimmed = end;
  while (trimmed > start) {
    const lastLineStart = lineStart(source, trimmed - 1);
    if (source.slice(lastLineStart, trimmed).trim() !== '') {
      break;
    }
    trimmed = lastLineStart;
  }
  return trimmed;
}

/**
 * findDeviceFragment locates one device in the whole-config YAML by name.
 * Returns null when the config does not parse, has no `devices` list, or has
 * no device by that name — the caller falls back to whole-config editing
 * rather than showing a pane built on a guess.
 */
export function findDeviceFragment(configText: string, deviceName: string): DeviceFragment | null {
  const doc = parseDocument(configText);
  if (doc.errors.length > 0) {
    return null;
  }
  const devices = doc.get('devices');
  if (!isSeq(devices)) {
    return null;
  }

  for (const [index, item] of devices.items.entries()) {
    if (!isMap(item) || item.get('name') !== deviceName) {
      continue;
    }
    const range = (item as Node).range;
    if (!range) {
      return null;
    }
    const [nodeStart, valueEnd] = range;
    const start = lineStart(configText, nodeStart);

    // Bounded by where the next device starts, not by scanning forward from
    // this one's value-end. A block map's value-end can already sit on the
    // following line, and scanning from there swallowed the next device's
    // `- name:` line whenever the two were adjacent -- which is how the
    // daemon writes every config it saves, since yaml.Marshal puts no blank
    // line between sequence items. Splicing that fragment back deleted the
    // following device and grafted its fields onto this one.
    const nextStart = nextItemStart(configText, devices.items[index + 1]);
    const end =
      nextStart === null
        ? blockEnd(configText, start, valueEnd)
        : trimTrailingBlankLines(configText, start, nextStart);
    const block = configText.slice(start, end);

    // The first line carries the `- ` marker; the rest are indented to line up
    // past it. Both come off so the pane shows a device, not a list item.
    const markerMatch = /^(\s*)-\s+/.exec(block);
    if (!markerMatch) {
      return null;
    }
    const indent = markerMatch[1] ?? '';
    const bodyIndent = ' '.repeat(markerMatch[0].length);
    const text = block
      .split('\n')
      .map((line, index) =>
        index === 0 ? line.slice(markerMatch[0].length) : stripPrefix(line, bodyIndent),
      )
      .join('\n');

    return { text, start, end, indent };
  }
  return null;
}

/** Removes prefix from line when present; blank lines pass through. */
function stripPrefix(line: string, prefix: string): string {
  return line.startsWith(prefix) ? line.slice(prefix.length) : line;
}

/**
 * spliceDeviceFragment puts an edited device back into the whole config,
 * re-indented to the depth the list uses. Every byte outside the fragment's
 * range is preserved exactly, including comments and the spacing between
 * devices.
 */
export function spliceDeviceFragment(
  configText: string,
  fragment: DeviceFragment,
  replacement: string,
): string {
  const marker = `${fragment.indent}- `;
  const bodyIndent = ' '.repeat(marker.length);
  const lines = replacement.replace(/\n$/, '').split('\n');
  const block = lines
    .map((line, index) => {
      if (line === '') {
        return '';
      }
      return index === 0 ? marker + line : bodyIndent + line;
    })
    .join('\n');

  return `${configText.slice(0, fragment.start)}${block}\n${configText.slice(fragment.end)}`;
}

File: internal/config/device_fragment.go
package config

import (
	"strings"

	"gopkg.in/yaml.v3"
)

// DeviceFragment is one device's byte range inside a whole-config YAML
// document. The detail pane edits a single device, but the file belongs to the
// operator: re-serialising the whole config to persist one edit reformats it
// and drops the comments they wrote, so the edit is put back as a byte range
// instead — everything outside [Start, End) is copied through untouched.
//
// This mirrors ui/src/utils/device-fragment.ts, which did the same splice in
// the browser while a per-device save still went through the whole-config
// endpoint. That endpoint is now admin-only (#2173), so the splice moved here.
type DeviceFragment struct {
	// Start is the offset of the first byte of the device's `- ` line.
	Start int
	// End is the offset just past the device's block.
	End int
	// Indent is the whitespace the list items sit at, restored on splice.
	Indent string
}

// FindDeviceFragment locates one device in a whole-config YAML document by
// name. It reports false when the document does not parse, has no `devices`
// sequence, or has no device by that name — the caller then falls back to
// re-serialising rather than splicing on a guess.
func FindDeviceFragment(source, deviceName string) (DeviceFragment, bool) {
	devices, ok := deviceSequence(source)
	if !ok {
		return DeviceFragment{}, false
	}

	lines := strings.SplitAfter(source, "\n")
	offsets := lineOffsets(lines)

	for _, item := range devices {
		if item.Kind != yaml.MappingNode || mappingValue(item, "name") != deviceName {
			continue
		}
		// item.Line is 1-based and points at the `- name:` line, because a
		// block mapping's first key shares that line with the sequence marker.
		if item.Line < 1 || item.Line > len(lines) {
			return DeviceFragment{}, false
		}

		start := offsets[item.Line-1]

		indent, marked := markerIndent(lines[item.Line-1])
		if !marked {
			return DeviceFragment{}, false
		}

		end := blockEnd(lines, offsets, item.Line-1, len(indent))

		return DeviceFragment{Start: start, End: end, Indent: indent}, true
	}

	return DeviceFragment{}, false
}

// SpliceDeviceFragment puts an edited device back into the whole config,
// re-indented to the depth the list uses. Every byte outside the fragment's
// range is preserved exactly, including comments and the spacing between
// devices.
func SpliceDeviceFragment(source string, fragment DeviceFragment, replacement string) string {
	marker := fragment.Indent + "- "
	bodyIndent := strings.Repeat(" ", len(marker))

	var block strings.Builder

	for i, line := range strings.Split(strings.TrimSuffix(replacement, "\n"), "\n") {
		if i > 0 {
			block.WriteString("\n")
		}

		if line == "" {
			continue
		}

		if i == 0 {
			block.WriteString(marker)
		} else {
			block.WriteString(bodyIndent)
		}

		block.WriteString(line)
	}

	return source[:fragment.Start] + block.String() + "\n" + source[fragment.End:]
}

// deviceSequence returns the items of the top-level `devices` sequence.
func deviceSequence(source string) ([]*yaml.Node, bool) {
	var document yaml.Node
	if err := yaml.Unmarshal([]byte(source), &document); err != nil {
		return nil, false
	}

	if document.Kind != yaml.DocumentNode || len(document.Content) == 0 {
		return nil, false
	}

	root := document.Content[0]
	if root.Kind != yaml.MappingNode {
		return nil, false
	}

	for i := 0; i+1 < len(root.Content); i += 2 {
		if root.Content[i].Value != "devices" {
			continue
		}

		value := root.Content[i+1]
		if value.Kind != yaml.SequenceNode {
			return nil, false
		}

		return value.Content, true
	}

	return nil, false
}

// blockEnd walks forward from the device's `- ` line to the first line that
// leaves the item — a non-blank line indented no deeper than the marker, which
// is the next device, a comment written at the list's own level, or a dedented
// key. Trailing blank lines are then given back to the gap between devices, so
// editing one device does not close up the spacing around it.
func blockEnd(lines []string, offsets []int, markerIndex, indent int) int {
	end := len(lines)

	for i := markerIndex + 1; i < len(lines); i++ {
		if strings.TrimSpace(lines[i]) == "" {
			continue
		}

		if lineIndent(lines[i]) <= indent {
			end = i

			break
		}
	}

	for end > markerIndex+1 && strings.TrimSpace(lines[end-1]) == "" {
		end--
	}

	if end >= len(lines) {
		return offsets[len(offsets)-1]
	}

	return offsets[end]
}

// markerIndent returns the whitespace before a sequence item's `- ` marker.
func markerIndent(line string) (string, bool) {
	trimmed := strings.TrimLeft(line, " ")
	if !strings.HasPrefix(trimmed, "- ") {
		return "", false
	}

	return line[:len(line)-len(trimmed)], true
}

// lineIndent counts the leading spaces of a line.
func lineIndent(line string) int {
	return len(line) - len(strings.TrimLeft(line, " "))
}

// lineOffsets returns the byte offset each line starts at, plus a final entry
// for the end of the document so callers can address "just past the last line".
func lineOffsets(lines []string) []int {
	offsets := make([]int, len(lines)+1)

	total := 0
	for i, line := range lines {
		offsets[i] = total
		total += len(line)
	}

	offsets[len(lines)] = total

	return offsets
}

// mappingValue returns a mapping's scalar value for key, or "".
func mappingValue(mapping *yaml.Node, key string) string {
	for i := 0; i+1 < len(mapping.Content); i += 2 {
		if mapping.Content[i].Value == key {
			return mapping.Content[i+1].Value
		}
	}

	return ""
}
